flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AsyncKeyedProcessOperator.java

```diff
@@ -0,0 +1,220 @@
+
+/** A {@link StreamOperator} for executing {@link AsyncKeyedProcessFunction}. */
+@Internal
+public class AsyncKeyedProcessOperator<K, IN, OUT>
+        extends AbstractAsyncStateUdfStreamOperator<OUT, AsyncKeyedProcessFunction<K, IN, OUT>>
+        implements OneInputStreamOperator<IN, OUT>, Triggerable<K, VoidNamespace> {
+
+    private static final long serialVersionUID = 1L;
+
+    // Shared timestamp variable for collector, context and onTimerContext.
+    private transient DeclaredVariable<Long> sharedTimestamp;
+
+    private transient TimestampedCollectorWithDeclaredVariable<OUT> collector;
+
+    private transient ContextImpl context;
+
+    private transient OnTimerContextImpl onTimerContext;
+
+    private transient DeclarationContext declarationContext;
+
+    private transient ThrowingConsumer<IN, Exception> processor;
+    private transient ThrowingConsumer<Long, Exception> timerProcessor;
+
+    public AsyncKeyedProcessOperator(AsyncKeyedProcessFunction<K, IN, OUT> function) {
+        super(function);
+    }
+
+    @Override
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public void open() throws Exception {
+        super.open();
+        declarationContext = new DeclarationContext(getDeclarationManager());
+        userFunction.open(declarationContext);
+        sharedTimestamp =
+                declarationContext.declareVariable(
+                        LongSerializer.INSTANCE,
+                        "_AsyncKeyedProcessOperator$sharedTimestamp",
+                        null);
+
+        collector = new TimestampedCollectorWithDeclaredVariable<>(output, sharedTimestamp);
+
+        InternalTimerService<VoidNamespace> internalTimerService =
+                getInternalTimerService("user-timers", VoidNamespaceSerializer.INSTANCE, this);
+
+        TimerService timerService = new SimpleTimerService(internalTimerService);
+
+        context = new ContextImpl(userFunction, timerService, sharedTimestamp);
+        onTimerContext =
+                new OnTimerContextImpl(
+                        userFunction, timerService, declarationContext, sharedTimestamp);
+
+        processor = userFunction.declareProcess(declarationContext, context, collector);
+        timerProcessor = userFunction.declareOnTimer(declarationContext, onTimerContext, collector);
+    }
+
+    @Override
+    public void onEventTime(InternalTimer<K, VoidNamespace> timer) throws Exception {
+        collector.setAbsoluteTimestamp(timer.getTimestamp());
+        invokeUserFunction(TimeDomain.EVENT_TIME, timer);
+    }
+
+    @Override
+    public void onProcessingTime(InternalTimer<K, VoidNamespace> timer) throws Exception {
+        collector.eraseTimestamp();
+        invokeUserFunction(TimeDomain.PROCESSING_TIME, timer);
+    }
+
+    @Override
+    public void processElement(StreamRecord<IN> element) throws Exception {
+        collector.setTimestamp(element);
+        processor.accept(element.getValue());
```

Comment:
Do all other async operators have to use `declaring interface`?
If they don't use `declaring interface`, can they work normally?

Reply:
If they use original `KeyedProcessFunction`, it works.

This is a new code path for declare.